### DISCUSSION TITLE
Method to remove Topic(s) directly from trained/merged model based on Topic ID or Name?

### DISCUSSION THREAD

**Discussion by shuanglovesdata** [MaartenGr/BERTopic#2172] [Q&A]
Hi, I am working on a use case where I want to incrementally fit a Bertopic model with HDBSCAN over time when new data come in. I found that the [merge model](https://maartengr.github.io/BERTopic/getting_started/merge/merge.html) method very useful for that. 

However, I also would like to periodically remove stale topics that no longer have documents coming in after a while. I tried to use the merge or update topic method suggested in this [issue](https://github.com/MaartenGr/BERTopic/issues/669), but since I am merging models over time, it is very hard to keep track of the docs that were used to create a particular topic after new topics fitted with new sets of documents were merged into that topic. 

Anyone ran into this use case as well? Is there a way to achieve removing Topics by its ID or name in merged models without bringing in the docs that these models were trained on?

**Comment by MaartenGr** [comment]
That's a good question. I don't think that is currently implemented as we typically want to recalculate the topic representations as we merge them with other topics or when we update the topic assignments themselves. That said, it would indeed be a nice feature to have. I'm thinking the most easy to implement way would be to adjust `merge_topics` a bit. 

It contains the following code:

https://github.com/MaartenGr/BERTopic/blob/4042e744a66c174da15979206376f377f906a08c/bertopic/_bertopic.py#L2144-L2151

Which at some places require the `documents` variable that contains a dataframe of documents and their respective topics. What I would propose is to "delete" a topic by simply merging it with the -1 but **without** recalculating the topic representations. That way, you do not need the actual content of the documents, just their assignments. 

I think it would require a couple of updates:

```python
# Update topics
documents.Topic = documents.Topic.map(mapping)
self.topic_mapper_.add_mappings(mapping, topic_model=self)
documents = self._sort_mappings_by_frequency(documents) 
# self._extract_topics(documents, mappings=mappings) <- SKIP THIS
self._update_topic_size(documents)
# self._save_representative_docs(documents) <- SKIP THIS
self.probabilities_ = self._map_probabilities(self.probabilities_)
```

So we would skip the recalculation of the topic representations and representative documents (which require the documents) and instead update the following variables and remove the topic (and potentially re-order everything else):
* self.c_tf_idf_
* self.topic_embeddings_
* self.topic_representations_
* self.ctfidf_model
* self.topic_aspects_

**Reply by shuanglovesdata** [reply]
Thank you for the quick response @MaartenGr! I'd love to help write and add a delete_topics method and contribute to the package if you could guide me through it. Based on what you said, it would probably look something like this:

    def delete_topics(
        self,
        # docs: List[str],
        # topics_to_merge: List[Union[Iterable[int], int]],
        topics_to_delete: List[int],
        # images: List[str] = None,
    ) -> None:
        """
        docstring...
        """
        check_is_fitted(self)
        # check_documents_type(docs)
    
        topics_to_merge = [-1].extend(topics_to_delete)

        # or would a bettwer way to do this be building a list of [-1, X] like:
        # topics_to_merge = [[-1, x],
        #                     [-1, y].....]

        documents = pd.DataFrame(
            {
                # "Document": docs,
                "Topic": self.topics_,
                # "Image": images,
                # "ID": range(len(docs)),
            }
        )
    
        mapping = {topic: topic for topic in set(self.topics_)}
        if isinstance(topics_to_merge[0], int):
            for topic in sorted(topics_to_merge):
                mapping[topic] = topics_to_merge[0]
        elif isinstance(topics_to_merge[0], Iterable):
            for topic_group in sorted(topics_to_merge):
                for topic in topic_group:
                    mapping[topic] = topic_group[0]
        else:
            raise ValueError(
                "Make sure that `topics_to_merge` is either" "a list of topics or a list of list of topics."
            )
    
        # Track mappings and sizes of topics for merging topic embeddings
        mappings = defaultdict(list)
        for key, val in sorted(mapping.items()):
            mappings[val].append(key)
        mappings = {
            topic_to: {
                "topics_from": topics_from,
                "topic_sizes": [self.topic_sizes_[topic] for topic in topics_from],
            }
            for topic_to, topics_from in mappings.items()
        }
    
        # Update topics
        documents.Topic = documents.Topic.map(mapping)
        self.topic_mapper_.add_mappings(mapping, topic_model=self)
        documents = self._sort_mappings_by_frequency(documents)
        # self._extract_topics(documents, mappings=mappings)
        self._update_topic_size(documents)
        # self._save_representative_docs(documents)
        self.probabilities_ = self._map_probabilities(self.probabilities_)
    
        #TODOS: update the following for topics_to_delete:
        # self.c_tf_idf_
        # self.topic_embeddings_
        # self.topic_representations_
        # self.ctfidf_model
        # self.topic_aspects_


I do also have some conceptual questions, please forgive my shallow understanding of the inner-working of the code base:

1. Is the -1 topic used at inference/transform time in some way? Would keep merging valid topics/clusters into -1 cause more documents to get assigned to the -1 topic in the future or prevent similar future clusters of those merged into -1 to form again?

2. What would be the best way to keep track of the remaining/new topics after deletion/merging some into -1, considering I have an ongoing process of merging models sequentially over time. Based on the examples in [merge model](https://maartengr.github.io/BERTopic/getting_started/merge/merge.html), without deletion or merging topics into -1, it looks like the combination of Topic (ID) and Name should be unique, and one should never encounter topics with the same ID+Name over time. However, considering deletion or merging into -1,  is there a need to come up with a way to make sure a future cluster do not take the same ID/Name of one that was deleted?

**Comment by MaartenGr** [comment]
```python
# or would a bettwer way to do this be building a list of [-1, X] like:
# topics_to_merge = [[-1, x],
#                     [-1, y].....]
```

I think I prefer the other example, simply add all topics to -1 in a single list if `merge_topics` already supports that behavior. 

```python
documents = pd.DataFrame(
    {
        # "Document": docs,
        "Topic": self.topics_,
        # "Image": images,
        # "ID": range(len(docs)),
    }
)
```

Note that we might still need those columns, but they can be empty strings.

```python
    #TODOS: update the following for topics_to_delete:
    # self.c_tf_idf_
    # self.topic_embeddings_
    # self.topic_representations_
    # self.ctfidf_model
    # self.topic_aspects_
```

Also note that when you update those variables, you only need to remove the selected topics. For ctfidf, ctfidf model (idf), topic embeddings, that only means removing the vectors/values for that topic. For all others, it means removing the selected topics and reducing the topic ids of the remaining topics. 

For instance, in `self.topic_representations_`, you might delete topic 2 and as a result should reduce the topic ids of topics 3-5 to 2-4 as there cannot be missing topic ids. 
 
All of this is not trivial, so it might require a bit of testing to get everything right. I'll try to help out wherever I can.

**Comment by shuanglovesdata** [comment]
Hi @MaartenGr, apologies for not following up on this for so long. I finally got around to this and think it would still be useful to have a way to delete topic without the need for accumulating docs. I gave it a try based on what we talked about previously. Could you please take a look at this [branch](https://github.com/shuanglovesdata/BERTopic-SC/tree/add-delete-topics) at the forked repo? I tried to adapt test_merge to test_delete, but it is giving me a lot of errors (due to clustering models that do not give -1 cluster and some mapping logic). I did my own tests in dev_test_sc.ipynb which looks fine with an HDBSCAN example. Would love to get your take on how to modify the method so it will work on various types of models/use cases. Ultimately, I just would like a way to delete topics without docs. For my use case of keep merging models over time, it is not essential to track mapping all the way back to the original or even the last merge of the models. It is more important that after deleting a topic, it allows for the creation of a highly similar but different topic in the future. Thanks!

**Reply by MaartenGr** [reply]
> I tried to adapt test_merge to test_delete, but it is giving me a lot of errors (due to clustering models that do not give -1 cluster and some mapping logic)

Could you tell me a bit more about the errors that are popping up? Also, does this mean that it does work for HDBSCAN?

I can imagine it to be more tricky with cluster models that do not have -1 values, considering you are essentially then creating a new topic (-1). So, things like probability matrices need to be updated. 

> Ultimately, I just would like a way to delete topics without docs. For my use case of keep merging models over time, it is not essential to track mapping all the way back to the original or even the last merge of the models.

For the package itself, the `delete_topic` function should update the mapping, considering this function might also be used for other use cases. However, if this is working for your use case, then you could consider not pushing a PR for this package. Although it is highly appreciated to have these kinds of functionalities, I wouldn't want to burden you too much with things that are outside of your scope (and time!).

**Comment by shuanglovesdata** [comment]
HI @MaartenGr,

Thanks for the quick response! I do still want to contribute to this awesome package if possible. After re-reading some of your previous comments, I made some updates to the delete_topics method by using as many components of merge_topics as I can, and I also updated the test to account for models without -1 topics. Now it passes all the tests including those in test_delete.py.

I am not sure if the way I am handling the new -1 topic attributes aligns with you, so could you please review and let me know your thoughts? Happy to iterate on further.

Here is the branch: https://github.com/shuanglovesdata/BERTopic-SC/tree/add-delete-topics

**Reply by MaartenGr** [reply]
Could you perhaps create a PR for this? It would make it easier for me to review.

**Reply by shuanglovesdata** [reply]
Sure, here it is: https://github.com/MaartenGr/BERTopic/pull/2322

**Comment by shuanglovesdata** [comment]
@MaartenGr, following up on this PR. Any feedback on how it should be improved? Thank you!

**Reply by shuanglovesdata** [reply]
Sorry, not sure why i did not get a notification for your comments on the PR. I will go through those and reply. Thanks for reviewing!

**Comment by shuanglovesdata** [comment]
Hi @MaartenGr, thank you for reviewing the PR! I resolved the comments. Please let me know if further updates are needed.

**Comment by shuanglovesdata** [comment]
Hi @MaartenGr, I replied to your comment on the PR and the new commit fixes the issues you mentioned. I also ran the test case you provided and the output looks fine. Could you please double check and  let me know if further updates are needed? Thank you.

**Comment by shuanglovesdata** [comment]
@MaartenGr, thanks for getting back to me on my pull request: https://github.com/MaartenGr/BERTopic/pull/2322. I could not replicate your issue, could you please take a look at my comments there with screenshots and code, and double-check?
